File: examples/ts/code/v1alpha/image_to_image.ts
import fetch from 'node-fetch'
import FormData from 'form-data'
import fs from 'node:fs'

const engineId = 'stable-diffusion-v1-5'
const apiHost = process.env.API_HOST ?? 'https://api.stability.ai'
const apiKey = process.env.STABILITY_API_KEY

if (!apiKey) throw new Error('Missing Stability API key.')

// NOTE: This example is using a NodeJS FormData library. Browser
// implementations should use their native FormData class. React Native
// implementations should also use their native FormData class.
const formData = new FormData()
formData.append('init_image', fs.readFileSync('../init_image.png'))
formData.append(
  'options',
  JSON.stringify({
    cfg_scale: 7,
    clip_guidance_preset: 'FAST_BLUE',
    step_schedule_start: 0.6,
    step_schedule_end: 0.01,
    height: 512,
    width: 512,
    samples: 1,
    steps: 50,
    text_prompts: [
      {
        text: 'Galactic dog',
        weight: 1,
      },
    ],
  })
)

const response = await fetch(
  `${apiHost}/v1alpha/generation/${engineId}/image-to-image`,
  {
    method: 'POST',
    headers: {
      ...formData.getHeaders(),
      Accept: 'image/png',
      Authorization: apiKey,
    },
    body: formData,
  }
)

if (!response.ok) {
  throw new Error(`Non-200 response: ${await response.text()}`)
}

try {
  const writeStream = fs.createWriteStream(`./out/v1alpha_img2img.png`)
  response.body?.pipe(writeStream)
} catch (e) {
  console.error(e)
}

File: examples/go/code/v1beta/image_to_image.go
package main

import (
	"bytes"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"mime/multipart"
	"net/http"
	"os"
)

type ImageToImageImage struct {
	Base64       string `json:"base64"`
	Seed         uint32 `json:"seed"`
	FinishReason string `json:"finishReason"`
}

type ImageToImageResponse struct {
	Images []ImageToImageImage `json:"artifacts"`
}

func main() {
	engineId := "stable-diffusion-v1-5"

	// Build REST endpoint URL
	apiHost, hasApiHost := os.LookupEnv("API_HOST")
	if !hasApiHost {
		apiHost = "https://api.stability.ai"
	}
	reqUrl := apiHost + "/v1beta/generation/" + engineId + "/image-to-image"

	// Acquire an API key from the environment
	apiKey, hasAPIKey := os.LookupEnv("STABILITY_API_KEY")
	if !hasAPIKey {
		panic("Missing STABILITY_API_KEY environment variable")
	}

	data := &bytes.Buffer{}
	writer := multipart.NewWriter(data)

	// Write the init image to the request
	initImageWriter, _ := writer.CreateFormField("init_image")
	initImageFile, initImageErr := os.Open("../init_image.png")
	if initImageErr != nil {
		panic("Could not open init_image.png")
	}
	_, _ = io.Copy(initImageWriter, initImageFile)

	// Write the options to the request
	_ = writer.WriteField("init_image_mode", "IMAGE_STRENGTH")
	_ = writer.WriteField("image_strength", "0.35")
	_ = writer.WriteField("text_prompts[0][text]", "Galactic dog with a cape")
	_ = writer.WriteField("cfg_scale", "7")
	_ = writer.WriteField("clip_guidance_preset", "FAST_BLUE")
	_ = writer.WriteField("height", "512")
	_ = writer.WriteField("width", "512")
	_ = writer.WriteField("samples", "1")
	_ = writer.WriteField("steps", "50")
	writer.Close()

	// Execute the request
	payload := bytes.NewReader(data.Bytes())
	req, _ := http.NewRequest("POST", reqUrl, payload)
	req.Header.Add("Content-Type", writer.FormDataContentType())
	req.Header.Add("Accept", "application/json")
	req.Header.Add("Authorization", "Bearer "+apiKey)
	res, _ := http.DefaultClient.Do(req)
	defer res.Body.Close()

	if res.StatusCode != 200 {
		var body map[string]interface{}
		if err := json.NewDecoder(res.Body).Decode(&body); err != nil {
			panic(err)
		}
		panic(fmt.Sprintf("Non-200 response: %s", body))
	}

	// Decode the JSON body
	var body ImageToImageResponse
	if err := json.NewDecoder(res.Body).Decode(&body); err != nil {
		panic(err)
	}

	// Write the images to disk
	for i, image := range body.Images {
		outFile := fmt.Sprintf("./out/v1beta_img2img_%d.png", i)
		file, err := os.Create(outFile)
		if err != nil {
			panic(err)
		}

		imageBytes, err := base64.StdEncoding.DecodeString(image.Base64)
		if err != nil {
			panic(err)
		}

		if _, err := file.Write(imageBytes); err != nil {
			panic(err)
		}

		if err := file.Close(); err != nil {
			panic(err)
		}
	}
}
